protos/core.proto

```diff
@@ -260,127 +250,422 @@ message CompiledGraph {
+    // setting applies to all partitions in a table, but is calculated
+    // independently for each partition based on the partition time.
+    int32 partition_expiration_days = 5;
+
+    // Declares whether the partitioned table requires a WHERE clause predicate
+    // filter that filters the partitioning column.
+    bool require_partition_filter = 6;
+
+    // Key-value pairs for the table, view, and materialized view options.
+    map<string, string> additional_options = 7;
   }
-}
 
-message CompileExecutionRequest {
-  CompileConfig compile_config = 1;
-}
+  message Table {
+    // Target is the unique action identifier.
+    Target target = 1;
 
-message CompileExecutionResponse {
-  CompiledGraph compiled_graph = 1;
-}
+    // Canonical Target is the unique action identifier, with no target
+    // overrides applied.
+    // This option cannot be manually specified in action configs.
+    Target canonical_target = 2;
 
-// This feature list is added to when making potentilly backwards breaking
-// changes. It lets the caller of Dataform Core know whether it supports the
-// change.
-enum SupportedFeatures {
-  UNKNOWN_FEATURE = 0;
-  ARRAY_BUFFER_IPC = 1;
-}
+    // Targets (non-canonical) of actions that this action is dependent on.
+    repeated Target dependency_targets = 3;
 
-message ActionConfigs {
-  repeated ActionConfig actions = 1;
-}
+    // Path to the source file that the contents of the action is loaded from.
+    string source = 4;
 
-message ActionConfig {
-  // Target is a unique action identifier.
-  Target target = 1;
+    // Path that the config of the action is loaded from.
+    // This option cannot be manually specified in action configs.
+    string config_path = 5;
 
-  // A list of user-defined tags with which the action should be labeled.
-  repeated string tags = 2;
+    // A list of user-defined tags with which the action should be labeled.
+    repeated string tags = 6;
 
-  // Targets of actions that this action is dependent on.
-  repeated Target dependency_targets = 3;
+    // If set to true, this action will not be executed. However, the action may
+    // still be depended upon.
+    bool disabled = 7;
 
-  // Filename to load the action contents from.
-  string file_name = 4;
+    // Describes a struct, object or record in a dataset that has nested
+    // columns.
+    ActionDescriptor action_descriptor = 8;
+
+    // The SQL query.
+    string query = 9;
 
-  // Type specific configs. Must correspond to the declared type.
-  oneof action_config {
-    TableConfig table = 5;
-    ViewConfig view = 6;
-    IncrementalTableConfig incremental_table = 7;
-    AssertionConfig assertion = 8;
-    OperationConfig operation = 9;
-    DeclarationConfig declaration = 10;
-    NotebookConfig notebook = 11;
+    // Pre/post operations.
+    repeated string pre_ops = 10;
+    repeated string post_ops = 11;
+
+    // Common BigQuery options.
+    BigQueryOptions bigquery = 12;
```

Comment:
tempting to flatten these. Or, I would at least rename them - they are really "table options".

Reply:
Flattening to the specific action types makes sense to me, particularly looking at logic like https://github.com/dataform-co/dataform/blob/f14c861488ea04e7862995aa8ee82a9657824867/core/session.ts#L615